In [3]:
# Dependencies and Setup
import pandas as pd
import numpy as np

# File to Load (Remember to Change These)
file_to_load = "Downloads/Resources_purchase_data.csv"
# Read Purchasing File and store into Pandas data frame
purchase_data = pd.read_csv(file_to_load)


In [2]:
purchase_data.columns

Index(['Purchase ID', 'SN', 'Age', 'Gender', 'Item ID', 'Item Name', 'Price'], dtype='object')

In [4]:
#use nunique to get unique count of players
unique_table=pd.DataFrame({
    "Total Players": [purchase_data["SN"].nunique()]})
unique_table

Total Players
0            576

In [5]:
#get unique count of ID
unique_item=purchase_data["Item ID"].nunique()

#calculate required field
average_price=purchase_data["Price"].mean()
number_purchase=purchase_data["Purchase ID"].count()
total_revenue=purchase_data["Price"].sum()

#convert to DataFrame
summary_table=pd.DataFrame({
    "Number of Unique Items": [unique_item],
    "Average Price": [average_price],
    "Number of Purchases": [number_purchase],
    "Total Revenue": [total_revenue]
})

#format columns
summary_table["Average Price"]=summary_table["Average Price"].map("${:.2f}".format)
summary_table["Total Revenue"]=summary_table["Total Revenue"].map("${:.2f}".format)

#sort column in required order
summary_table=summary_table[["Number of Unique Items", "Average Price", "Number of Purchases", "Total Revenue"]]
summary_table

Number of Unique Items Average Price  Number of Purchases Total Revenue
0                     183         $3.05                  780      $2379.77

In [6]:
#get two needed columns
unique_SN=purchase_data[["SN","Gender"]]
#get unique SN
unique_SN=unique_SN.drop_duplicates("SN")
#get gender statistic
gender_count=unique_SN["Gender"].value_counts()
#get total unique count of SN
unique=unique_SN["SN"].nunique()

#convert to dataframe
gender_frame=pd.DataFrame(gender_count)

#get percentage of each gender
percentage_of_players=gender_frame["Gender"]/unique*100
#add percentage column to dataframe
gender_frame["Percentage of Players"]=percentage_of_players
#format percentage
gender_frame["Percentage of Players"]=gender_frame["Percentage of Players"].map("{:.2f}".format)
gender_frame

Gender Percentage of Players
Male                      484                 84.03
Female                     81                 14.06
Other / Non-Disclosed      11                  1.91

In [7]:
#get needed columns
purchase_analysis=purchase_data[["Gender","Price"]]

#groupby gender
gendergroup=purchase_analysis.groupby("Gender")

#convert groupby into Dataframe
purchase_analysis= pd.DataFrame(gendergroup.count())

#get required fields from groupby
pricemean=pd.DataFrame(gendergroup.mean())
pricetotal=pd.DataFrame(gendergroup.sum())

#add fields to dataframe
purchase_analysis["Average Purchase Pirce"]=pricemean
purchase_analysis["Total Purchase Value"]=pricetotal
price_mean_total=purchase_analysis["Total Purchase Value"]/gender_count
purchase_analysis["Avg Total Purchase per Person"]=price_mean_total

#format columns
purchase_analysis["Average Purchase Pirce"]=purchase_analysis["Average Purchase Pirce"].map("${:.2f}".format)
purchase_analysis["Total Purchase Value"]=purchase_analysis["Total Purchase Value"].map("${:.2f}".format)
purchase_analysis["Avg Total Purchase per Person"]=purchase_analysis["Avg Total Purchase per Person"].map("${:.2f}".format)

#rename columns
purchase_analysis.rename(columns={"Price": "Purchase Count"}, inplace=True)
purchase_analysis

Purchase Count Average Purchase Pirce  \
Gender                                                         
Female                            113                  $3.20   
Male                              652                  $3.02   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase Value Avg Total Purchase per Person  
Gender                                                                    
Female                             $361.94                         $4.47  
Male                              $1967.64                         $4.07  
Other / Non-Disclosed               $50.19                         $4.56

In [8]:
#create bins and group
bins=[0,9,14,19,24,29,34,39,np.inf]
groups=["<10","10-14","15-19","20-24","25-29","30-34","35-39","40+"]

#remove duplicates in data
unique_age=purchase_data.drop_duplicates("SN")

#turn dataset into bins
unique_age["Total Count"]=pd.cut(unique_age["Age"], bins, labels=groups)

#get count by age group
age_group_count=unique_age["Total Count"].value_counts(sort=False)

#turn result into dataframe
age_demography=pd.DataFrame(age_group_count)

#get statistic and add to dataframe and then format
age_percentage=age_demography["Total Count"]/unique*100
age_demography["Percentage of Players"]=age_percentage
age_demography["Percentage of Players"]=age_demography["Percentage of Players"].map("{:.2f}".format)

#sort age group, let "<10" display on the first row
age_demography["new"]=range(1,len(age_demography)+1)
age_demography.loc["<10", "new"]=0
age_demography_sort=age_demography.sort_values("new").drop("new", axis=1)

age_demography_sort

C:\Users\james\Anaconda3\lib\site-packages\ipykernel_launcher.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if __name__ == '__main__':


Total Count Percentage of Players
<10             17                  2.95
10-14           22                  3.82
15-19          107                 18.58
20-24          258                 44.79
25-29           77                 13.37
30-34           52                  9.03
35-39           31                  5.38
40+             12                  2.08

In [9]:
#cut data into bins to age group from last question
purchase_data["Purchase Count"]=pd.cut(purchase_data["Age"], bins,labels=groups )

#get count number from each age group and turn into dataframe
purchase_count=purchase_data["Purchase Count"].value_counts(sort=False)
purchase_value=purchase_data.groupby("Purchase Count")
purchase_analysis_df=pd.DataFrame(purchase_count)

#get required statistics and add to columns to dataframe
purchase_avg_df=pd.DataFrame(purchase_value.mean())
purchase_analysis_df["Average Purchase Price"]= purchase_avg_df["Price"]
purchase_value_df=pd.DataFrame(purchase_value.sum())
purchase_analysis_df["Total Purchase Value"]= purchase_value_df["Price"]
avg_total_value=purchase_analysis_df["Total Purchase Value"]/age_group_count
purchase_analysis_df["Avg Total Purchase per Person"]=avg_total_value

#format columns
purchase_analysis_df["Average Purchase Price"]=purchase_analysis_df["Average Purchase Price"].map("${:.2f}".format)
purchase_analysis_df["Total Purchase Value"]=purchase_analysis_df["Total Purchase Value"].map("${:.2f}".format)
purchase_analysis_df["Avg Total Purchase per Person"]=purchase_analysis_df["Avg Total Purchase per Person"].map("${:.2f}".format)

purchase_analysis_df

Purchase Count Average Purchase Price Total Purchase Value  \
<10                23                  $3.35               $77.13   
10-14              28                  $2.96               $82.78   
15-19             136                  $3.04              $412.89   
20-24             365                  $3.05             $1114.06   
25-29             101                  $2.90              $293.00   
30-34              73                  $2.93              $214.00   
35-39              41                  $3.60              $147.67   
40+                13                  $2.94               $38.24   

      Avg Total Purchase per Person  
<10                           $4.54  
10-14                         $3.76  
15-19                         $3.86  
20-24                         $4.32  
25-29                         $3.81  
30-34                         $4.12  
35-39                         $4.76  
40+                           $3.19

In [10]:
#create groupby data, create required statistics using agg function
spender=purchase_data.groupby("SN").agg({"Price": {"Purchase Count": "count",
                                                   "Average Purchase Price": "mean",
                                                   "Total Purchase Price":sum
                                                   }})

#remove top level Price in order to show result as required
spender.columns=spender.columns.droplevel(0)

#sort data and display top 5
spender_top=spender.sort_values(["Total Purchase Price"], ascending=False).head(5)

#format columns
spender_top["Average Purchase Price"]=spender_top["Average Purchase Price"].map("${:.2f}".format)
spender_top["Total Purchase Price"]=spender_top["Total Purchase Price"].map("${:.2f}".format)

spender_top

C:\Users\james\Anaconda3\lib\site-packages\pandas\core\groupby\groupby.py:4656: FutureWarning: using a dict with renaming is deprecated and will be removed in a future version
  return super(DataFrameGroupBy, self).aggregate(arg, *args, **kwargs)


Purchase Count Average Purchase Price Total Purchase Price
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

In [11]:
#create groupby data, create required statistics using agg function
popular_item=purchase_data.groupby(["Item ID","Item Name"]).agg({"Price":{"Purchase Count":"count",
                                                                         "Item Price": "mean",
                                                                         "Total Purchase Value":sum}})
#remove top level Price in order to show result as required
popular_item.columns=popular_item.columns.droplevel(0)

#sort data and display top 5
most_popular_item=popular_item.sort_values(["Purchase Count"],ascending=False).head(5)

#format columns
most_popular_item["Item Price"]=most_popular_item["Item Price"].map("${:.2f}".format)
most_popular_item["Total Purchase Value"]=most_popular_item["Total Purchase Value"].map("${:.2f}".format)
most_popular_item

C:\Users\james\Anaconda3\lib\site-packages\pandas\core\groupby\groupby.py:4656: FutureWarning: using a dict with renaming is deprecated and will be removed in a future version
  return super(DataFrameGroupBy, self).aggregate(arg, *args, **kwargs)


Purchase Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
108     Extraction, Quickblade Of Trembling Hands                  9   
82      Nirvana                                                    9   
19      Pursuit, Cudgel of Necromancy                              8   

                                                     Item Price  \
Item ID Item Name                                                 
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
108     Extraction, Quickblade Of Trembling Hands         $3.53   
82      Nirvana                                           $4.90   
19      Pursuit, Cudgel of Necromancy                     $1.02   

                                                     Total Purchase Value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
108     Extraction, Quickblade Of Trembling Hands                  $31.77  
82      Nirvana                                                    $44.10  
19      Pursuit, Cudgel of Necromancy                               $8.16

In [12]:
#sort data and display top 5
most_profitable_item=popular_item.sort_values(["Total Purchase Value"], ascending=False).head(5)

#format columns
most_profitable_item["Item Price"]=most_profitable_item["Item Price"].map("${:.2f}".format)
most_profitable_item["Total Purchase Value"]=most_profitable_item["Total Purchase Value"].map("${:.2f}".format)
most_profitable_item

Purchase Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
92      Final Critic                                               8   
103     Singed Scalpel                                             8   

                                                     Item Price  \
Item ID Item Name                                                 
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
92      Final Critic                                      $4.88   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
92      Final Critic                                               $39.04  
103     Singed Scalpel                                             $34.80

Week 4 HW - Heroes of Pymoli
                        3 Observable trends based on data:

1)Based on the data, a very large amount of males played the game Heroes of Pymoli. 84.03% of players were male to be exact.

2) Although fewer female players, they had a tendency to spend more money on the game than males. Female players on average spent $3.20 compared to males at $3.02. 

3)The largest demographic age range of players is 20-24. That age range making up 44% of all players. 
